dig/graph.go

```diff
@@ -183,5 +183,38 @@ func (g *graph) registerConstructor(c interface{}) error {
 	}
 
 	g.nodes[objType] = &n
+
+	// object needs to be part of the graph to properly detect cycles
+	if cycleErr := g.detectCycles(&n); cycleErr != nil {
+		// if the cycle was detected delete from the graph
+		delete(g.nodes, objType)
```

Comment:
Register can be called at runtime though. If we stick to our philosophy of not calling panic outside of service startup, this should return error.